### PR TITLE
Add username validation feature

### PR DESCRIPTION
**Pull request by Smnrgcl**
This commit introduces a new validation rule in the service layer to ensure that usernames do not contain any special characters, enhancing the system's security and data integrity. It also includes comprehensive unit tests for the new validation logic, covering various edge cases to ensure robustness and reliability.

### CODE REVIEW

**Comment by Sddilora** [comment]
Enhancements are needed in error handling as the message 'Username is not valid' lacks specificity, potentially confusing users about the rules and necessary corrections.

**Comment by Sddilora** [comment]
There are the definitions of the Error variables in, [domain package](https://github.com/octoposprime/op-be-user/blob/phase1/internal/domain/model/object/error.go). 

```go
        ErrorUserUsernameIsEmpty                            error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrEmpty)
	ErrorUserUsernameIsExists                           error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrAlreadyExists)
	ErrorUserUsernameIsTooShort                         error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooShort)
	ErrorUserUsernameIsTooLong                          error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooLong)
	ErrorUserUsernameIsNotValid                         error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrNotValid)
```
"Username cannot contain special characters" needs to be implemented in this way too. You can find the smdol.Errbase or other declarations in [model package of the shared library](https://github.com/octoposprime/op-be-shared/blob/phase1/pkg/model/error.go)

**Review by Sddilora — CHANGES_REQUESTED** [review]
Error handling methods should adhere to the project structure and be enhanced. 
❗❗❗Please add the IDs of the issue that will be closed and resolved with this pull request. ( To close the issue testFunctions for those functions need to be implemented too #5 )

**Review by husamettinarabaci — CHANGES_REQUESTED** [review]
We have struct-based error types. Please keep our generic mo.Error*** styles. You have to add all new errors in model/object/error.go.

Note: in the Golang; any error message shouldn't start with a capital character.

File: internal/domain/service/service_test.go [hunk 1]
```diff
@@ -1,103 +1,73 @@
-package domain_test
+ppackage domain
```
Comment (husamettinarabaci) [inline internal/domain/service/service_test.go]: Please check this

**Review by husamettinarabaci — CHANGES_REQUESTED** [review]
Please caution this message;
https://github.com/octoposprime/op-be-user/pull/16#issuecomment-1975302539

**Comment by husamettinarabaci** [comment]
Please remove;

var (
    ErrorUserUsernameIsEmpty            = errors.New("username is empty")
    ErrorUserUsernameIsTooShort         = errors.New("username is too short")
    ErrorUserUsernameIsTooLong          = errors.New("username is too long")
    ErrorUserUsernameContainsSpace      = errors.New("username contains space")
    ErrorUserUsernameContainsSpecialChars = errors.New("username contains special characters")
    ErrorUserUsernameContainsInvalidChars = errors.New("username contains invalid characters")
)

**Comment by Sddilora** [comment]
image
As in Godev's example, the name of the test suite should be the same as the name of the source package (the name of the package containing the function under test).

Also, such things should not be changed :)

File: internal/domain/service/service_test.go [hunk 2]
```diff
@@ -1,29 +1,34 @@
-package domain
```
Comment (Sddilora) [inline internal/domain/service/service_test.go:1]: This part should be remained as "domain".

File: internal/domain/service/service_test.go [hunk 3]
```diff
@@ -1,29 +1,34 @@
...
+package domain_test
```
Comment (Sddilora) [inline internal/domain/service/service_test.go]: This part should be remained as "domain".

File: internal/domain/model/object/error.go [hunk 4]
```diff
@@ -67,6 +71,8 @@ var (
 	ErrorUserUsernameIsExists                           error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrAlreadyExists)
 	ErrorUserUsernameIsTooShort                         error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooShort)
 	ErrorUserUsernameIsTooLong                          error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooLong)
+	ErrorUserUsernameContainsInValidChars               error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrContainsInvalidChars)
+	ErrorUserUsernameContainsSpecialChars               error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrorUserUsernameContainsSpecialChars )
```
Comment (Sddilora) [inline internal/domain/model/object/error.go]: initialization cycle: ErrorUserUsernameContainsSpecialChars refers to itselfcompiler[InvalidInitCycle](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#InvalidInitCycle)

File: internal/domain/service/service.go [hunk 5]
```diff
@@ -1,93 +1,140 @@
...
+    }
+    if len(user.UserName) < 8 {
+        return mo.ErrorUserUsernameIsTooShort
+    }
+    if len(user.UserName) > 20 {
+        return mo.ErrorUserUsernameIsTooLong
+    }
+    // Check for spaces
+    if strings.Contains(user.UserName, " ") {
+        return mo.ErrorUserUsernameContainsSpace
```
Comment (Sddilora) [inline internal/domain/service/service.go]: this is not present in mo package

File: internal/domain/model/object/error.go [hunk 6]
```diff
@@ -67,6 +71,8 @@ var (
 	ErrorUserUsernameIsExists                           error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrAlreadyExists)
 	ErrorUserUsernameIsTooShort                         error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooShort)
 	ErrorUserUsernameIsTooLong                          error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooLong)
+	ErrorUserUsernameContainsInValidChars               error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrContainsInvalidChars)
```
Comment (Sddilora) [inline internal/domain/model/object/error.go]: typo causes error

**Comment by Sddilora** [comment]
Let's start fresh with a new branch and close this pull request, as the current branch and PR are causing some confusion.
